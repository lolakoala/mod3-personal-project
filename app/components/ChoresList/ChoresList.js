import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

class ChoresList extends Component {
  getUserChores = (chores, id, which) => {
    const userChores = chores.filter(chore => chore.assignedTo === id);
    const userChoresToDo = userChores.filter(chore => chore.done === false)
      .sort((itemA, itemB) => Date.parse(itemA.datePosted) - Date.parse(itemB.datePosted));
    const userChoresDone = userChores.filter(chore => chore.done === true)
      .sort((itemA, itemB) => Date.parse(itemA.datePosted) - Date.parse(itemB.datePosted));
    return which === 'all' ? [...userChoresToDo, ...userChoresDone] : userChoresToDo;
  }

  getMatchingChores = (chores, value) => {
    return chores.filter(chore => {
      return chore.title.includes(value) || chore.details.includes(value);
    }).sort((itemA, itemB) => Date.parse(itemB.datePosted) - Date.parse(itemA.datePosted));
  }

  getChoresToMap = () => {
    const { usersHouse, placeRendered, currentUser, searchValue } = this.props;
    if (placeRendered === '/houselist') {
      return usersHouse.chores.sort((itemA, itemB) => Date.parse(itemB.datePosted) - Date.parse(itemA.datePosted));
    } else if (placeRendered === '/userlist') {
      return this.getUserChores(usersHouse.chores, currentUser.id, 'all');
    } else if (placeRendered === '/') {
      return this.getUserChores(usersHouse.chores, currentUser.id, 'todo');
    } else if (placeRendered === 'summary') {
      return usersHouse.chores.filter(chore => chore.done === false).sort((itemA, itemB) => Date.parse(itemB.datePosted) - Date.parse(itemA.datePosted));
    } else if (placeRendered === 'search') {
      return this.getMatchingChores(usersHouse.chores, searchValue);
    }
  }

  getError = () => {
    const { placeRendered, usersHouse } = this.props;
    if (placeRendered === '/') {
      return 'You have chores to do.';
    } else if (placeRendered === '/houselist') {
      return `${usersHouse.houseName} has no chores posted.`;
    } else if (placeRendered === 'summary') {
      return `${usersHouse.houseName} has no chores to do.`;
    } else if (placeRendered === 'search') {
      return 'No chores match your query.';
    }
  }

  claimChore = (userId, chore, usersHouse) => {
    this.props.claimChore(userId, chore, usersHouse);
    this.forceUpdate();
  }

  markChoreDone = (chore, usersHouse) => {
    this.props.markChoreDone(chore, usersHouse);
    this.forceUpdate();
  }

  render() {
    const { placeRendered, usersHouse, currentUser } = this.props;
    if (usersHouse.chores.length && usersHouse.chores[0].title !== 'fake') {
      const choresToMap = this.getChoresToMap();
      if (choresToMap.length) {
        return (
          <div className='choreslist'>
            {placeRendered === '/' ? <h4>Chores I Need To Do</h4> : null}
            {placeRendered === 'summary' ? <h4>Chores To Be Done</h4> : null}
            {choresToMap.map(chore => {
              return (<div key={chore.datePosted} className='chore'>
                <Link to={`chores/${chore.id}`}>{chore.title}</Link>
                <p>{chore.urgency === 'high' ? 'urgent' : 'not urgent'}</p>
                <p className='claim-button' onClick={() => this.claimChore(currentUser.id, chore, usersHouse)}>
                  {chore.assignedTo.length ? 'Claimed' : 'Claim Chore'}
                </p>
                <p className='choredone' onClick={() => this.markChoreDone(chore, usersHouse)}>{chore.done === false ? 'Mark Done' : 'Done'}</p>
              </div>);
            })}
          </div>
        );
      } else {
        return <p className='error'>{this.getError()}</p>;
      }
    } else {
      return <p className='error'>{`${usersHouse.houseName} has no chores posted.`}</p>;
    }
  }
}

export default ChoresList;

ChoresList.propTypes = {
  usersHouse: PropTypes.object,
  currentUser: PropTypes.object,
  placeRendered: PropTypes.string,
  markChoreDone: PropTypes.func,
  claimChore: PropTypes.func,
  searchValue: PropTypes.string
};
