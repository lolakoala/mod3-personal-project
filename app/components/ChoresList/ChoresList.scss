.choreslist h4,
.error {
  background-color: rgba(141, 132, 73, 0.7);
  color: #3d6aa2;
  font-family: 'Shadows Into Light', cursive;
  font-size: 20px;
  width: 250px;
  padding: 5px;
  text-align: center;
  height: 35px;
  border-radius: 5px;
  line-height: 35px;
  margin-top: 10px;
}

.choreslist .chore {
  display: flex;
  height: 30px;
  width: 300px;
  border-radius: 5px;
  border: 2px solid rgba(141, 132, 73, 0.7);
  margin-top: 10px;
  line-height: 30px;
  // text-align: center;
  font-family: 'BenchNine', sans-serif;
  position: relative;
}

.chore a {
  text-decoration: none;
  width: 75px;
  overflow: hidden;
  display: inline-block;
  margin: 2px 5px;
  background-color: #D83D59;
  border: none;
  // font-family: 'BenchNine', sans-serif;
  height: 25px;
  color: #F4CCE2;
  font-size: 20px;
  border-radius: 25px;
  text-align: center;
}

.chore a:hover {
  cursor: pointer;
  border: 2px dashed #D83D59;
  background-color: #F4CCE2;
  color: #D83D59;
}

.choredone, .claim-button {
  height: 100%;
   width: 75px;
   background-color: #eb6b3c;
   position: absolute;
   top: 0;
   right: 0;
   color: #F4CCE2;
   display: inline-block;
   font-size: 15px;
   padding-left: 10px;
}

.choredone:hover,
.claim-button:hover {
    cursor: pointer;
    background-color: #F4CCE2;
    color: #eb6b3c;
}

.claim-button {
  right: 90px;
  width: 50px;
}
// .chore .due,
// .chore .total {
//   width: 60px;
//   font-size: 15px;
//   display: inline-block;
//   margin: 2px 0px 2px 5px;
//   text-align: center;
// }
//
// #done-button {
//   height: 100%;
//   width: 75px;
//   background-color: #eb6b3c;
//   position: absolute;
//   top: 0;
//   right: 0;
//   color: #F4CCE2;
//   display: inline-block;
//   font-size: 15px;
//   padding-left: 10px;
// }

// .done-button:hover {
//   cursor: pointer;
//   background-color: #F4CCE2;
//   color: #eb6b3c;
// }
//
