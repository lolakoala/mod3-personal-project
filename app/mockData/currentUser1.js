const currentUser1 = {
  id : "RVde1m9BhkQBLZlTJNsFlQZXqP43",
  name: "Amanda Brenner",
  house: "Dog9"
};

export default currentUser1;
