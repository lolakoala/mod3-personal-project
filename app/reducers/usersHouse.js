import firebase from '../firebase.js';

const usersHouse = (state = {}, action) => {
  const housesRef = firebase.database().ref('houses');
  switch (action.type) {
  case 'CREATE_HOUSE':
    housesRef.push(action.house);
    return action.house;
  case 'GET_HOUSE':
    const updatedHouse = Object.assign(action.usersHouse, { users: [...action.usersHouse.users, action.user] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updatedHouse);
    return updatedHouse;
  case 'LOGIN_SUCCESS':
    return action.usersHouse;
  case 'SIGNOUT':
    return {};
  case 'LEAVE_HOUSE':
    const deleteUser = action.usersHouse.users.filter(user => user.id !== action.currentUser.id);
    const houseWithoutUser = Object.assign(action.usersHouse, { users: deleteUser });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithoutUser);
    return {};
  case 'ADD_BILL':
    const takeOutFakeBills = action.usersHouse.bills.filter(bill => bill.title !== 'fake');
    const houseWithBill = Object.assign(action.usersHouse, { bills: [...takeOutFakeBills, action.bill] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithBill);
    return houseWithBill;
  case 'ADD_BULLETIN':
    const takeOutFakeBulletins = action.usersHouse.bulletins.filter(bulletin => bulletin.title !== 'fake');
    const houseWithBulletin = Object.assign(action.usersHouse, { bulletins: [...takeOutFakeBulletins, action.bulletin] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithBulletin);
    return houseWithBulletin;
  case 'ADD_CHORE':
    const takeOutFakeChore = action.usersHouse.chores.filter(chore => chore.title !== 'fake');
    const houseWithChore = Object.assign(action.usersHouse, { chores: [...takeOutFakeChore, action.chore] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithChore);
    return houseWithChore;
  case 'ADD_READER':
    const oldBulletin = action.usersHouse.bulletins.find(bulletin => bulletin.id === action.bulletinId);
    if (oldBulletin.hasRead.includes(action.userId)) {
      return action.usersHouse;
    }
    const newBulletin = Object.assign(oldBulletin, { hasRead: [...oldBulletin.hasRead, action.userId] });
    const updateBulletins = action.usersHouse.bulletins.filter(bulletin => {
      return bulletin.id !== action.bulletinId;
    });
    const updateWithBulletin = Object.assign(action.usersHouse, { bulletins: [...updateBulletins, newBulletin] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updateWithBulletin);
    return updateWithBulletin;
  case 'MARK_PAID':
    const oldBill = action.usersHouse.bills.find(bill => bill.id === action.billId);
    const user = oldBill.allUsersTotals.find(user => user.id === action.userId);
    if (user.paid === true) {
      return action.usersHouse;
    }
    const newUser = Object.assign(user, { paid: true });
    const newAllUsersTotals = oldBill.allUsersTotals.filter(user => user.id !== action.userId);
    const newBill = Object.assign(oldBill, { allUsersTotals: [...newAllUsersTotals, newUser] });
    const updatedBills = action.usersHouse.bills.filter(bill => {
      return bill.id !== action.billId;
    });
    const updateWithBill = Object.assign(action.usersHouse, { bills: [...updatedBills, newBill] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updateWithBill);
    return updateWithBill;
  default:
    return state;
  }
};

export default usersHouse;
