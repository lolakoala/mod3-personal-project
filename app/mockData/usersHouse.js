import currentUser1 from './currentUser1.js';
import currentUser2 from './currentUser2.js';

const bill1 = {
  allUsersTotals: [
    {
      id: currentUser1.id,
      paid: false,
      total: 50
    },
    {
      id: currentUser2.id,
      paid: false,
      total: 50
    }
  ],
  datePosted: "11/01/2017",
  details: "We used a lot of water.",
  duedate: "11/9/17",
  id: "b3019762705779",
  parsedDuedate: 1510210800000,
  postedBy: currentUser1,
  title: "Water Bill",
  total: "100"
};

const bulletin1 = {
  datePosted: "11/01/2017",
  details: "The dogs got out through a hole in the fence yesterday.",
  hasRead: [currentUser1.id],
  id: 1509567098747,
  postedBy: currentUser1,
  title: "Don't leave dogs out unattended."
};

const chore1 = {
  assignedTo: currentUser1.id,
  datePosted: "10/31/2017",
  details: "Empty the dishwasher.",
  done: false,
  id: "c1509485241076",
  postedBy: currentUser1,
  title: "Dishes",
  urgency: "low"
};

const usersHouse = {
  houseKey: "-KxoIBWy5UYRscFYdWty",
  bills: [bill1],
  bulletins: [bulletin1],
  chores: [chore1],
  houseCode: "woof9",
  houseName: "Dog9",
  users: [currentUser1, currentUser2]
};

export default usersHouse;
