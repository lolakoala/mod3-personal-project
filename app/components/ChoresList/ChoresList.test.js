import React from "react";
import ChoresList from './ChoresList.js';
import { shallow } from "enzyme";
import usersHouse from '../../mockData/usersHouse.js';
import currentUser1 from '../../mockData/currentUser1.js';

describe('ChoresList component', () => {
  const mockFn=jest.fn();
  const wrapper = shallow(<ChoresList usersHouse={usersHouse}
    currentUser={currentUser1}
    placeRendered={'/houselist'}
    markChoreDone={mockFn}
    claimChore={mockFn}
    searchValue={''}/>);

  it('should exist', () => {
    expect(wrapper).toBeDefined();
  });

  it('should have a title', () => {
    const title = wrapper.find('.choreslist-title');

    expect(title.text()).toEqual('House Chores');
  });

  it('should have a title column header', () => {
    const titleHeader = wrapper.find('.title-header');

    expect(titleHeader.text()).toEqual('Title');
  });

  it('should have an urgency column header', () => {
    const urgencyHeader = wrapper.find('.urgency-header');

    expect(urgencyHeader.text()).toEqual('Urgency');
  });

  it('should have a claimed header', () => {
    const claimedHeader = wrapper.find('.claimed-header');

    expect(claimedHeader.text()).toEqual('Claimed by');
  });

  it('should have a done header', () => {
    const doneHeader = wrapper.find('.done-header');

    expect(doneHeader.text()).toEqual('Done');
  });

  it('should have display all chores of the house', () => {
    const chores = wrapper.find('.chore');

    expect(chores.length).toEqual(usersHouse.chores.length);
  });

  it('should fire action', () => {
    const claimButton = wrapper.find('.claim-button').first();
    const doneButton = wrapper.find('.done-button').first();

    claimButton.simulate('click');
    doneButton.simulate('click');
    expect(mockFn).toHaveBeenCalledTimes(2);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
