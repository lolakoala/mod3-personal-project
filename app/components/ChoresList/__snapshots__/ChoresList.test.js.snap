// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChoresList component should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ChoresList
    claimChore={[Function]}
    currentUser={
      Object {
        "house": "Dog9",
        "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
        "name": "Amanda Brenner",
      }
    }
    markChoreDone={[Function]}
    placeRendered="/houselist"
    searchValue=""
    usersHouse={
      Object {
        "bills": Array [
          Object {
            "allUsersTotals": Array [
              Object {
                "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
                "paid": false,
                "total": 50,
              },
              Object {
                "id": "cFNdorl5ceRyMr9XGvPFRVEcw7I3",
                "paid": false,
                "total": 50,
              },
            ],
            "datePosted": "11/01/2017",
            "details": "We used a lot of water.",
            "duedate": "11/9/17",
            "id": "b3019762705779",
            "parsedDuedate": 1510210800000,
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Water Bill",
            "total": "100",
          },
        ],
        "bulletins": Array [
          Object {
            "datePosted": "11/01/2017",
            "details": "The dogs got out through a hole in the fence yesterday.",
            "hasRead": Array [
              "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            ],
            "id": 1509567098747,
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Don't leave dogs out unattended.",
          },
        ],
        "chores": Array [
          Object {
            "assignedTo": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            "datePosted": "10/31/2017",
            "details": "Empty the dishwasher.",
            "done": false,
            "id": "c1509485241076",
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Dishes",
            "urgency": "low",
          },
        ],
        "houseCode": "woof9",
        "houseKey": "-KxoIBWy5UYRscFYdWty",
        "houseName": "Dog9",
        "users": Array [
          Object {
            "house": "Dog9",
            "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            "name": "Amanda Brenner",
          },
          Object {
            "house": "Dog9",
            "id": "cFNdorl5ceRyMr9XGvPFRVEcw7I3",
            "name": "Lola Koala",
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        null,
        null,
        Array [
          <div
            className="chore"
          >
            <Link
              replace={false}
              to="chores/c1509485241076"
            >
              Dishes
            </Link>
            <p>
              not urgent
            </p>
            <p
              className="claim-button"
              onClick={[Function]}
            >
              Claimed
            </p>
            <p
              className="choredone"
              onClick={[Function]}
            >
              Mark Done
            </p>
          </div>,
        ],
      ],
      "className": "choreslist",
    },
    "ref": null,
    "rendered": Array [
      null,
      null,
      Object {
        "instance": null,
        "key": "10/31/2017",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Link
              replace={false}
              to="chores/c1509485241076"
            >
              Dishes
            </Link>,
            <p>
              not urgent
            </p>,
            <p
              className="claim-button"
              onClick={[Function]}
            >
              Claimed
            </p>,
            <p
              className="choredone"
              onClick={[Function]}
            >
              Mark Done
            </p>,
          ],
          "className": "chore",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "class",
            "props": Object {
              "children": "Dishes",
              "replace": false,
              "to": "chores/c1509485241076",
            },
            "ref": null,
            "rendered": "Dishes",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "not urgent",
            },
            "ref": null,
            "rendered": "not urgent",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Claimed",
              "className": "claim-button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Claimed",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Mark Done",
              "className": "choredone",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Mark Done",
            "type": "p",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          null,
          null,
          Array [
            <div
              className="chore"
            >
              <Link
                replace={false}
                to="chores/c1509485241076"
              >
                Dishes
              </Link>
              <p>
                not urgent
              </p>
              <p
                className="claim-button"
                onClick={[Function]}
              >
                Claimed
              </p>
              <p
                className="choredone"
                onClick={[Function]}
              >
                Mark Done
              </p>
            </div>,
          ],
        ],
        "className": "choreslist",
      },
      "ref": null,
      "rendered": Array [
        null,
        null,
        Object {
          "instance": null,
          "key": "10/31/2017",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                replace={false}
                to="chores/c1509485241076"
              >
                Dishes
              </Link>,
              <p>
                not urgent
              </p>,
              <p
                className="claim-button"
                onClick={[Function]}
              >
                Claimed
              </p>,
              <p
                className="choredone"
                onClick={[Function]}
              >
                Mark Done
              </p>,
            ],
            "className": "chore",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "class",
              "props": Object {
                "children": "Dishes",
                "replace": false,
                "to": "chores/c1509485241076",
              },
              "ref": null,
              "rendered": "Dishes",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "not urgent",
              },
              "ref": null,
              "rendered": "not urgent",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Claimed",
                "className": "claim-button",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Claimed",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Mark Done",
                "className": "choredone",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Mark Done",
              "type": "p",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
